racecar_serial/pb2roscpp/src/pb2roscpp.cpp

```diff
@@ -0,0 +1,298 @@
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Float32MultiArray>> publisher_;
+    std::shared_ptr<rclcpp::Subscription<geometry_msgs::msg::Twist>> subscription_;
+
+    serial::Serial
+        *my_serialp; // Serial pointer to initialize the serial communication in the constructor
+
+    struct message { // message structure for nanopb custom functions found on stackoverflow
+        int id;
+        std::string msg;
+    };
+
+    // Fonctions
+    void controlCallback(const geometry_msgs::msg::Twist msg)
+    {
+        // declare variables
+        uint8_t buffer[MAX_MSG_LEN];
+
+        // prepare the actual "variable" array
+        FloatList actualData = {};
+        FloatList_add_number(&actualData, (float)msg.angular.z);
+        FloatList_add_number(&actualData, (float)msg.linear.x);
+        FloatList_add_number(&actualData, (float)msg.linear.z);
+
+        // init the message
+        FloatArray message = FloatArray_init_zero;
+
+        // put the data in the message
+        message.data.arg = &actualData;
+
+        // link the encode function for float arrays to the message object
+        message.data.funcs.encode = FloatArray_encode_numbers;
+
+        /* Create a stream that will write to our buffer. */
+        pb_ostream_t ostream = pb_ostream_from_buffer(buffer, sizeof(buffer));
+
+        // encode the message
+        if (!pb_encode(&ostream, FloatArray_fields, &message)) {
+            const char *error = PB_GET_ERROR(&ostream);
+            printf("pb_encode error: %s\n", error);
+            return;
+        }
+
+        // Define vars for stringifing the encoded buffer
+        size_t total_bytes_encoded = ostream.bytes_written;
+        char toSend[10];
+        std::string strmsg;
+        std::string begining = "<1|";
+        std::string end = ";>";
+
+        // stringify
+        for (int j = 0; j < (int)total_bytes_encoded; ++j) {
+            sprintf(toSend, "%02X", buffer[j]);
+            strmsg += toSend;
+        }
+        // strmsg+='\0';
+        // build the message format
+        begining.append(strmsg);
+        begining.append(end);
+
+        // send the message
+
+        // std::cout<<(char*)begining.c_str()<<std::endl;
+        my_serialp->write((char *)begining.c_str());
+    }
+
+    void readMessage() // reading arduino messages callback
+    {
+        std::string msgbuffer;
+        int msgids[MAX_NUMBERS];
+        char *msgs[MAX_MSG_LEN];
+        uint8_t bufferin[MAX_MSG_LEN];
+        // empty array for decoding
+        FloatList decodedData = {};
+        /* Allocate space for the decoded message. */
+        FloatArray message = FloatArray_init_zero;
+        message.data.arg = &decodedData;
+        message.data.funcs.decode = FloatArray_decode_single_number;
+
+        if (my_serialp->available() > 0) { // wait for 1000 characters to enter the serialbuffer
+
+            my_serialp->readline(msgbuffer, MAX_MSG_LEN, ">"); // readuntil the end of a message
```

Comment:
This kind of unpacking is not reliable, if my msg containt the value the uint_8 '>' this will stop before the end. Normaly a state machine is used to unpack message that have header. Reading the start bytes (<1|) than reading the packet size as the data and finaly reading the stop bytes (>) then returning only the data